src/Widgets/TestCaseEdit.cpp

```diff
@@ -65,19 +80,64 @@ void TestCaseEdit::dropEvent(QDropEvent *event)
 {
     LOG_INFO("Object dropped into testcase widget");
     auto urls = event->mimeData()->urls();
-    if (!isReadOnly() && !urls.isEmpty())
+    if (role != Output && !urls.isEmpty())
     {
         LOG_INFO("Dropped file is " << urls[0].toLocalFile());
         loadFromFile(urls[0].toLocalFile());
         event->acceptProposedAction();
     }
 }
 
-void TestCaseEdit::modifyText(const QString &text)
+void TestCaseEdit::modifyText(const QString &text, bool keepHistory)
 {
-    auto cursor = textCursor();
-    cursor.select(QTextCursor::Document);
-    cursor.insertText(text);
+    this->text = text;
+
+    const int limit =
+        role == Output ? SettingsHelper::getOutputDisplayLengthLimit() : SettingsHelper::getLoadTestCaseLengthLimit();
+
+    QString displayText;
+
+    if (text.length() <= limit)
+    {
+        displayText = text;
+        if (role != Output)
+            setReadOnly(false);
+    }
+    else
+    {
+        LOG_INFO("Too long: " << INFO_OF(role) << INFO_OF(text.length()));
+
+        setReadOnly(true);
+
+        const QString name = role == Input ? tr("input") : (role == Output ? tr("output") : tr("expected"));
+        const QString setLimitPlace = role == Output ? tr("Preferences->Advanced->Limits->Output Display Length Limit")
+                                                     : tr("Preferences->Advanced->Limits->Load Test Case Length Limit");
+
+        displayText = tr("The %1 is too long, only the first %2 characters are shown. You can set the length limit in "
```

Comment:
Why do you think it should be prepend  instead of append?

Reply:
Because I want to let the user know this happens. Otherwise it's very possible that they don't know why the input is read-only.

Reply:
Or we can move this to the message logger.

Reply:
> Or we can move this to the message logger.

Yes, that will be better because showing long texts in Input/Output is not good ideas as user needs to scroll to read it.

Reply:
> user needs to scroll to read it

I think the user can simply ignore it and only have to read it once, but if it's in the mssage logger, the user needs to scroll to read other messages.

Reply:
Why not set shorten the message to something like: 
"Showing only the first 200 characters"
<200 characters of stdout>

In message logger, you can say in a warning:
"[Limits Warnings] : stdout exceeded maximum output limit set in Preferences->Advanced->Limits->Output Limit"

Reply:
Do you mean splitting the message into two parts, one in the TestCaseEdit and another in the message logger?

Reply:
Yes

Reply:
Maybe we can move this to the message logger, and make the "where to set limit" messages optional in the settings.

Reply:
I don't think a new settings for this is a good idea, This way our settings will be bloated.

Reply:
The HTML title seems like a good solution.